File: LeetCode/c++/SortColors.cpp
#include <iostream>
#include <vector>

using namespace std;

void arrangeColors(vector<int>& nums, int start, int end,int value)
{
	for (int i = start; i < end; i++)
		nums[i] = value;
}

void sortColors(vector<int>& nums) 
{
	int zeroCnt = 0;
	int oneCnt = 0;
	int twoCnt = 0;

	//统计数组中012的个数
	for (int i = 0; i < nums.size(); i++)
	{
		if (nums[i] == 0)
			zeroCnt++;
		else if (nums[i] == 1)
			oneCnt++;
		else if (nums[i] == 2)
			twoCnt++;
	}

	cout << zeroCnt << " " << oneCnt << " " << twoCnt << endl;

	//按统计好的个数分配数组空间
	arrangeColors(nums, 0, zeroCnt, 0);
	arrangeColors(nums, zeroCnt, zeroCnt+oneCnt, 1);
	arrangeColors(nums, oneCnt+zeroCnt, oneCnt + zeroCnt+twoCnt, 2);
}

int main()
{
	vector<int> v;
	v.push_back(1);
	v.push_back(0);
	v.push_back(2);
	v.push_back(2);
	v.push_back(0);
	v.push_back(1);
	v.push_back(0);
	v.push_back(0);
	v.push_back(2);

	sortColors(v);

	for (int i = 0; i < v.size(); i++)
		cout << v[i] << " ";

	return 0;
}

File: LeetCode/iter1/python/SortColors.py
class Solution(object):
    def arrangeColors(self, nums,start,end,value):
        """
        :type nums: List[int]
        :type start: integer
        :type end: integer
        :type value: integer
        :rtype: void Do not return anything, modify nums in-place instead.
        """    
        for i in range(start,end,1):
            nums[i] = value     
    
    def sortColors(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        zeroCnt = 0
        oneCnt = 0
        twoCnt = 0
        #统计012出现的次数
        for i in nums:
            if(i == 0):
                zeroCnt += 1
            elif(i == 1):
                oneCnt += 1
            elif(i==2):
                twoCnt += 1
        
        #根据次数重新分配012
        self.arrangeColors(nums,0,zeroCnt,0)
        self.arrangeColors(nums,zeroCnt,zeroCnt+oneCnt,1)
        self.arrangeColors(nums,zeroCnt+oneCnt,zeroCnt+oneCnt+twoCnt,2)
        
#测试程序
s = Solution()
num = [0,1,1,1,0,0,0,0,2,2,2,2]
s.sortColors(num)
print(num)